### ISSUE TITLE
NICOS: Change Proxy to use Active MQ

### ISSUE THREAD

**Issue by DominicOram** [ISISComputingGroup/IBEX#1547]
Depends on #1456.

The proxy currently uses raw sockets for sending/receiving information from the client that wishes to connect to NICOS. Zero MQ should be used instead. As part of this ticket the required python libraries should be added into our python installation

**Comment by FreddieAkeroyd** [comment]
Just a random though, we are already using JMS / ActiveMQ within IBEX, another option?

**Comment by DominicOram** [comment]
At first glance I thought that JMS/ActiveMQ would be an annoying amount of overhead but actually it shouldn't be too bad as they are already used elsewhere.  Do you think it would be better to run up another activeMQ instance with the scriptserver or just use a separate topic/queue?

**Comment by DominicOram** [comment]
PR at https://github.com/ISISComputingGroup/nicos-core/pull/6

**Comment by John-Holt-Tessella** [comment]
Not sure hoe to test this I wanted to see a script running in the gui but couldn't. I think this is because I have not understood how it works.

**Comment by DominicOram** [comment]
Testing might be more clear after GUI chat on 11/11/16, otherwise I can talk you through it

**Comment by John-Holt-Tessella** [comment]
This sequence doesn't work and I don't know why. Is this a bug in nicos or the proxy?

Run stop script server and then start_script_server. Then using client:

```
which client do you want to send from: 
Connected to JMS server: tcp://localhost:61616
Connected to JMS server: tcp://localhost:61616
0
type some text to send: 
SCRIPT
which client do you want to send from: 
####
Message on client: 0
In response to: null
{"payload": {"custom_path": "C:\\Instrument\\Apps\\EPICS\\ISIS\\ScriptServer\\nicos-core\\master\\custom", "nicos_root": "C:\\Instrument\\Apps\\EPICS\\ISIS\\ScriptServer\\nicos-core\\master", "rsakey": "LS0tLS1CRUdJTiBSU0EgUFVCTElDIEtFWS0tLS0tCk1FZ0NRUUNyK1lQSjFPSjBBQjVtQUpXcU5K\nQmNraktodXdYYjJwdm91bnp3bXFqeDhGT1paWFVIMGE1cE9YTGYKdlRMWVNVV0FnM0pmK3VkT0VT\neTJuMjdOb05RSkFnTUJBQUU9Ci0tLS0tRU5EIFJTQSBQVUJMSUMgS0VZLS0tLS0K\n", "daemon_version": "vendor_c1bf2c4-65-gbf5a", "pw_hashing": "rsa,sha1", "protocol_version": 14, "custom_version": "vendor_c1bf2c4-65-gbf5a"}, "success": true}
####
0
type some text to send: 
SCRIPT
which client do you want to send from: 
0
type some text to send: 
SCRIPT
which client do you want to send from: 

which client do you want to send from: 
0
type some text to send: 
SCRIPT
which client do you want to send from: 
0
type some text to send: 
LOGIN
which client do you want to send from: 
0
type some text to send: 
SCRIPT
which client do you want to send from:
```

The log messages in nicos are:

```
[Fri Nov 18 14:21:52 2016] @@@ Restarting child "NICOSDAEMON"
[Fri Nov 18 14:21:52 2016] @@@    (as C:\Instrument\Apps\EPICS\ISIS\ScriptServer
\nicos-core\master\start_nicos_daemon_cmd.bat)
[Fri Nov 18 14:21:52 2016] @@@ The PID of new child "NICOSDAEMON" is: 10496
[Fri Nov 18 14:21:52 2016] @@@ @@@ @@@ @@@ @@@
[Fri Nov 18 14:21:53 2016]
[Fri Nov 18 14:21:53 2016] C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core
\master>REM @echo off
[Fri Nov 18 14:21:53 2016]
[Fri Nov 18 14:21:53 2016] C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core
\master>set MYDIRNICOS=C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core\mas
ter\
[Fri Nov 18 14:21:53 2016]
[Fri Nov 18 14:21:53 2016] C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core
\master>call C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core\master\..\..\
..\..\config_env_base.bat
[Fri Nov 18 14:21:54 2016] [14:21:54] loading setup 'daemon' (Setup for the Ibex
 execution daemon)
[Fri Nov 18 14:21:54 2016] [14:21:54] creating device 'Auth'...
[Fri Nov 18 14:21:54 2016] [14:21:54] creating device 'Daemon'...
[Fri Nov 18 14:21:54 2016] [14:21:54] setups loaded: daemon
[Fri Nov 18 14:21:54 2016] [14:21:54] Daemon    : NICOS daemon vvendor_c1bf2c4-6
5-gbf5a started, starting server on localhost
[Fri Nov 18 14:21:54 2016] [14:21:54] Daemon    : script_thread (re)started
[Fri Nov 18 14:21:54 2016] [14:21:54] Daemon    : processing request 1
[Fri Nov 18 14:21:54 2016] [14:21:54] loading setup 'system' (Ibex System setup)

[Fri Nov 18 14:21:54 2016] [14:21:54] importing module commands.basic...
[Fri Nov 18 14:21:54 2016] [14:21:54] importing module IbexInstrument.genie_comm
ands...
[Fri Nov 18 14:21:54 2016] [14:21:54] Daemon    : [handler #1] connection from [
'NDW1407.isis.cclrc.ac.uk', '127.0.0.1']
[Fri Nov 18 14:21:55 2016] [14:21:55] pv prefix NDW1407
[Fri Nov 18 14:21:55 2016] pv prefix NDW1407
[Fri Nov 18 14:21:55 2016] [14:21:55] THIS IS NDW1407! (test machine)
[Fri Nov 18 14:21:55 2016] THIS IS NDW1407! (test machine)
[Fri Nov 18 14:21:55 2016] [14:21:55] ndw1407 will use init_default
[Fri Nov 18 14:21:55 2016] ndw1407 will use init_default
[Fri Nov 18 14:21:55 2016] [14:21:55] PV prefix is TE:NDW1407:
[Fri Nov 18 14:21:55 2016] PV prefix is TE:NDW1407:
[Fri Nov 18 14:21:55 2016] [14:21:55] loading setup 'startup' (Ibex NICOS startu
p setup)
[Fri Nov 18 14:21:55 2016] [14:21:55] creating device 'Space' (The amount of fre
e space for storing data)...
[Fri Nov 18 14:21:55 2016] [14:21:55] setups loaded: startup
[Fri Nov 18 14:21:55 2016] [14:21:55] no cache present, switching to master anyw
ay
[Fri Nov 18 14:21:55 2016] [14:21:55] switched to master mode
[Fri Nov 18 14:22:40 2016] [14:22:40] Daemon    : ERROR: [handler #1] unhandled
exception - ValueError - too many values to unpack
[Fri Nov 18 14:22:40 2016] [14:22:40] Daemon    : [handler #1] handler unregiste
red
```

Script server says:

```
[Fri Nov 18 14:21:52 2016] @@@ Restarting child "SCRIPTSERVER"
[Fri Nov 18 14:21:52 2016] @@@    (as C:\Instrument\Apps\EPICS\ISIS\ScriptServer
\nicos-core\master\start_isis_ss_proxy_cmd.bat)
[Fri Nov 18 14:21:52 2016] @@@ The PID of new child "SCRIPTSERVER" is: 17216
[Fri Nov 18 14:21:52 2016] @@@ @@@ @@@ @@@ @@@
[Fri Nov 18 14:21:53 2016]
[Fri Nov 18 14:21:53 2016] C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core
\master>REM @echo off
[Fri Nov 18 14:21:53 2016]
[Fri Nov 18 14:21:53 2016] C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core
\master>set MYDIRNICOS=C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core\mas
ter\
[Fri Nov 18 14:21:53 2016]
[Fri Nov 18 14:21:53 2016] C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core
\master>call C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core\master\..\..\
..\..\config_env_base.bat
[Fri Nov 18 14:22:20 2016] INFO: New client connected on queue /remote-temp-queu
e/ID:NDW1407-54602-1479478931833-0:2:1
[Fri Nov 18 14:22:40 2016] INFO: Message received from NDW1407-54602-14794789318
33-0: {"command": "start", "name": "test", "code": "sleep(1000)"}
[Fri Nov 18 14:23:09 2016] INFO: Message received from NDW1407-54602-14794789318
33-0: {"command": "start", "name": "test", "code": "sleep(1000)"}
[Fri Nov 18 14:23:26 2016] INFO: Message received from NDW1407-54602-14794789318
33-0: {"command": "start", "name": "test", "code": "sleep(1000)"}
[Fri Nov 18 14:23:26 2016] MAJOR: Data cannot be sent:
[Fri Nov 18 14:23:34 2016] INFO: Message received from NDW1407-54602-14794789318
33-0: {"command": "authenticate", "login": "ibex", "passwd": "a2eed0a7fcb214a497
052435191b5264cca5b687", "display": "TEST"}
[Fri Nov 18 14:23:34 2016] MAJOR: Data cannot be sent:
[Fri Nov 18 14:23:41 2016] INFO: Message received from NDW1407-54602-14794789318
33-0: {"command": "start", "name": "test", "code": "sleep(1000)"}
[Fri Nov 18 14:23:41 2016] MAJOR: Data cannot be sent:
```

**Comment by John-Holt-Tessella** [comment]
Would I expect to get a response to:

```
type some text to send: 
LOGIN
which client do you want to send from: 
####
Message on client: 0
In response to: null
{"payload": {"user_level": 20}, "success": true}
####
hi
which client do you want to send from: 
0
type some text to send: 
hi
```

**Comment by John-Holt-Tessella** [comment]
If I log in as channel 01. the run a script all is good.
If I run as channel 1 straight after and run the script without a log in I get in the script server:

```
[Fri Nov 18 14:44:29 2016] MAJOR: Error receiving from NICOS: ☻
[Fri Nov 18 14:45:22 2016] INFO: Message received from NDW1407-54675-14794793881
72-0: {"command": "start", "name": "test_1", "code": "sleep(2)\nprint'hi'\nsleep
(2)"}
[Fri Nov 18 14:45:22 2016] MAJOR: Error receiving from NICOS: pop from empty lis
tMAJOR: Error receiving from NICOS: Python int too large to convert to C long
[Fri Nov 18 14:45:22 2016]
```

but nicos says:

```
[Fri Nov 18 14:45:22 2016] [14:45:22] Daemon    : processing request 4
[Fri Nov 18 14:45:22 2016] [14:45:22] sleeping for 2.0 seconds...
[Fri Nov 18 14:45:24 2016] [14:45:24] hi
[Fri Nov 18 14:45:24 2016] hi
[Fri Nov 18 14:45:24 2016] [14:45:24] sleeping for 2.0 seconds...
```
which is my script running! This is not ideal is nicos not dealing with the log in or is it ignoring it? We pressumably need security here.

**Comment by DominicOram** [comment]
That's odd cos you got a response the first time which looks correct. Potentially it may be that you sent commands in the wrong order as a login should be first. Can you try testing the following:

Send any message content from one client, you should receive a response with the password hashing etc, like you did there.

Then from the same client send in a login command, you should receive a response saying success.

From there try sending scripts and pausing/running

The proxy should still be adapted so that the login process is a little clearer though so feel free to put this into rework.

Get Outlook for Android<https://aka.ms/ghei36>



On Fri, Nov 18, 2016 at 2:29 PM +0000, "John Holt" <notifications@github.com<mailto:notifications@github.com>> wrote:


This sequence doesn't work and I don't know why. Is this a bug in nicos or the proxy?

Run stop script server and then start_script_server. Then using client:

which client do you want to send from:
Connected to JMS server: tcp://localhost:61616
Connected to JMS server: tcp://localhost:61616
0
type some text to send:
SCRIPT
which client do you want to send from:
####
Message on client: 0
In response to: null
{"payload": {"custom_path": "C:\\Instrument\\Apps\\EPICS\\ISIS\\ScriptServer\\nicos-core\\master\\custom", "nicos_root": "C:\\Instrument\\Apps\\EPICS\\ISIS\\ScriptServer\\nicos-core\\master", "rsakey": "LS0tLS1CRUdJTiBSU0EgUFVCTElDIEtFWS0tLS0tCk1FZ0NRUUNyK1lQSjFPSjBBQjVtQUpXcU5K\nQmNraktodXdYYjJwdm91bnp3bXFqeDhGT1paWFVIMGE1cE9YTGYKdlRMWVNVV0FnM0pmK3VkT0VT\neTJuMjdOb05RSkFnTUJBQUU9Ci0tLS0tRU5EIFJTQSBQVUJMSUMgS0VZLS0tLS0K\n", "daemon_version": "vendor_c1bf2c4-65-gbf5a", "pw_hashing": "rsa,sha1", "protocol_version": 14, "custom_version": "vendor_c1bf2c4-65-gbf5a"}, "success": true}
####
0
type some text to send:
SCRIPT
which client do you want to send from:
0
type some text to send:
SCRIPT
which client do you want to send from:

which client do you want to send from:
0
type some text to send:
SCRIPT
which client do you want to send from:
0
type some text to send:
LOGIN
which client do you want to send from:
0
type some text to send:
SCRIPT
which client do you want to send from:


The log messages in nicos are:

[Fri Nov 18 14:21:52 2016] @@@ Restarting child "NICOSDAEMON"
[Fri Nov 18 14:21:52 2016] @@@    (as C:\Instrument\Apps\EPICS\ISIS\ScriptServer
\nicos-core\master\start_nicos_daemon_cmd.bat)
[Fri Nov 18 14:21:52 2016] @@@ The PID of new child "NICOSDAEMON" is: 10496
[Fri Nov 18 14:21:52 2016] @@@ @@@ @@@ @@@ @@@
[Fri Nov 18 14:21:53 2016]
[Fri Nov 18 14:21:53 2016] C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core
\master>REM @echo off
[Fri Nov 18 14:21:53 2016]
[Fri Nov 18 14:21:53 2016] C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core
\master>set MYDIRNICOS=C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core\mas
ter\
[Fri Nov 18 14:21:53 2016]
[Fri Nov 18 14:21:53 2016] C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core
\master>call C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core\master\..\..\
..\..\config_env_base.bat
[Fri Nov 18 14:21:54 2016] [14:21:54] loading setup 'daemon' (Setup for the Ibex
 execution daemon)
[Fri Nov 18 14:21:54 2016] [14:21:54] creating device 'Auth'...
[Fri Nov 18 14:21:54 2016] [14:21:54] creating device 'Daemon'...
[Fri Nov 18 14:21:54 2016] [14:21:54] setups loaded: daemon
[Fri Nov 18 14:21:54 2016] [14:21:54] Daemon    : NICOS daemon vvendor_c1bf2c4-6
5-gbf5a started, starting server on localhost
[Fri Nov 18 14:21:54 2016] [14:21:54] Daemon    : script_thread (re)started
[Fri Nov 18 14:21:54 2016] [14:21:54] Daemon    : processing request 1
[Fri Nov 18 14:21:54 2016] [14:21:54] loading setup 'system' (Ibex System setup)

[Fri Nov 18 14:21:54 2016] [14:21:54] importing module commands.basic...
[Fri Nov 18 14:21:54 2016] [14:21:54] importing module IbexInstrument.genie_comm
ands...
[Fri Nov 18 14:21:54 2016] [14:21:54] Daemon    : [handler #1] connection from [
'NDW1407.isis.cclrc.ac.uk', '127.0.0.1']
[Fri Nov 18 14:21:55 2016] [14:21:55] pv prefix NDW1407
[Fri Nov 18 14:21:55 2016] pv prefix NDW1407
[Fri Nov 18 14:21:55 2016] [14:21:55] THIS IS NDW1407! (test machine)
[Fri Nov 18 14:21:55 2016] THIS IS NDW1407! (test machine)
[Fri Nov 18 14:21:55 2016] [14:21:55] ndw1407 will use init_default
[Fri Nov 18 14:21:55 2016] ndw1407 will use init_default
[Fri Nov 18 14:21:55 2016] [14:21:55] PV prefix is TE:NDW1407:
[Fri Nov 18 14:21:55 2016] PV prefix is TE:NDW1407:
[Fri Nov 18 14:21:55 2016] [14:21:55] loading setup 'startup' (Ibex NICOS startu
p setup)
[Fri Nov 18 14:21:55 2016] [14:21:55] creating device 'Space' (The amount of fre
e space for storing data)...
[Fri Nov 18 14:21:55 2016] [14:21:55] setups loaded: startup
[Fri Nov 18 14:21:55 2016] [14:21:55] no cache present, switching to master anyw
ay
[Fri Nov 18 14:21:55 2016] [14:21:55] switched to master mode
[Fri Nov 18 14:22:40 2016] [14:22:40] Daemon    : ERROR: [handler #1] unhandled
exception - ValueError - too many values to unpack
[Fri Nov 18 14:22:40 2016] [14:22:40] Daemon    : [handler #1] handler unregiste
red


Script server says:

[Fri Nov 18 14:21:52 2016] @@@ Restarting child "SCRIPTSERVER"
[Fri Nov 18 14:21:52 2016] @@@    (as C:\Instrument\Apps\EPICS\ISIS\ScriptServer
\nicos-core\master\start_isis_ss_proxy_cmd.bat)
[Fri Nov 18 14:21:52 2016] @@@ The PID of new child "SCRIPTSERVER" is: 17216
[Fri Nov 18 14:21:52 2016] @@@ @@@ @@@ @@@ @@@
[Fri Nov 18 14:21:53 2016]
[Fri Nov 18 14:21:53 2016] C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core
\master>REM @echo off
[Fri Nov 18 14:21:53 2016]
[Fri Nov 18 14:21:53 2016] C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core
\master>set MYDIRNICOS=C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core\mas
ter\
[Fri Nov 18 14:21:53 2016]
[Fri Nov 18 14:21:53 2016] C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core
\master>call C:\Instrument\Apps\EPICS\ISIS\ScriptServer\nicos-core\master\..\..\
..\..\config_env_base.bat
[Fri Nov 18 14:22:20 2016] INFO: New client connected on queue /remote-temp-queu
e/ID:NDW1407-54602-1479478931833-0:2:1
[Fri Nov 18 14:22:40 2016] INFO: Message received from NDW1407-54602-14794789318
33-0: {"command": "start", "name": "test", "code": "sleep(1000)"}
[Fri Nov 18 14:23:09 2016] INFO: Message received from NDW1407-54602-14794789318
33-0: {"command": "start", "name": "test", "code": "sleep(1000)"}
[Fri Nov 18 14:23:26 2016] INFO: Message received from NDW1407-54602-14794789318
33-0: {"command": "start", "name": "test", "code": "sleep(1000)"}
[Fri Nov 18 14:23:26 2016] MAJOR: Data cannot be sent:
[Fri Nov 18 14:23:34 2016] INFO: Message received from NDW1407-54602-14794789318
33-0: {"command": "authenticate", "login": "ibex", "passwd": "a2eed0a7fcb214a497
052435191b5264cca5b687", "display": "TEST"}
[Fri Nov 18 14:23:34 2016] MAJOR: Data cannot be sent:
[Fri Nov 18 14:23:41 2016] INFO: Message received from NDW1407-54602-14794789318
33-0: {"command": "start", "name": "test", "code": "sleep(1000)"}
[Fri Nov 18 14:23:41 2016] MAJOR: Data cannot be sent:


�

**Comment by DominicOram** [comment]
A message of invalid Json gives no response from the proxy other than logging an error. It would be useful for the proxy to send a proper response to the client but I would suggest doing this in another ticket as this is not actually breaking functionality of the proxy.

Get Outlook for Android<https://aka.ms/ghei36>



On Fri, Nov 18, 2016 at 2:32 PM +0000, "John Holt" <notifications@github.com<mailto:notifications@github.com>> wrote:


Would I expect to get a response to:

type some text to send:
LOGIN
which client do you want to send from:
####
Message on client: 0
In response to: null
{"payload": {"user_level": 20}, "success": true}
####
hi
which client do you want to send from:
0
type some text to send:
hi


�

**Comment by DominicOram** [comment]
Yes, this appears to be a security issue. I will fix before merging.

Get Outlook for Android<https://aka.ms/ghei36>



On Fri, Nov 18, 2016 at 2:48 PM +0000, "John Holt" <notifications@github.com<mailto:notifications@github.com>> wrote:


If I log in as channel 01. the run a script all is good.
If I run as channel 1 straight after and run the script without a log in I get in the script server:

[Fri Nov 18 14:44:29 2016] MAJOR: Error receiving from NICOS: ☻
[Fri Nov 18 14:45:22 2016] INFO: Message received from NDW1407-54675-14794793881
72-0: {"command": "start", "name": "test_1", "code": "sleep(2)\nprint'hi'\nsleep
(2)"}
[Fri Nov 18 14:45:22 2016] MAJOR: Error receiving from NICOS: pop from empty lis
tMAJOR: Error receiving from NICOS: Python int too large to convert to C long
[Fri Nov 18 14:45:22 2016]


but nicos says:

[Fri Nov 18 14:45:22 2016] [14:45:22] Daemon    : processing request 4
[Fri Nov 18 14:45:22 2016] [14:45:22] sleeping for 2.0 seconds...
[Fri Nov 18 14:45:24 2016] [14:45:24] hi
[Fri Nov 18 14:45:24 2016] hi
[Fri Nov 18 14:45:24 2016] [14:45:24] sleeping for 2.0 seconds...


which is my script running! This is not ideal is nicos not dealing with the log in or is it ignoring it? We pressumably need security here.

**Comment by DominicOram** [comment]
To answer your issues in order:
1. I have modified the proxy so that the first command you send to it should be LOGIN.
2. I have created a new ticket https://github.com/ISISComputingGroup/IBEX/issues/1837 (also will help with issue 1)
3. This should be fixed, it was a bug in multiple clients.

Also, now see:
https://github.com/ISISComputingGroup/EPICS-IocLogServer/pull/8
